File: elixir/anagram/anagram.exs
defmodule Anagram do
  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t, [String.t]) :: [String.t]
  def match(base, candidates) do
    Enum.filter(candidates, &(is_different?(&1,base)))
    |> Enum.reduce([], fn word, acc -> acc ++ [%{sort_word(word) => word}] end)
    |> Enum.reduce([], fn mapped_candidate, acc -> acc ++ matches(sort_word(base), mapped_candidate) end)

  end
  
  #return matches
  def matches(string,map) do
    cond do
      map[string] -> [map[string]]
      true -> []
    end
  end
  
  def sort_word(string) do
    String.split(String.downcase(string),"", trim: true)
    |> Enum.sort
    |> List.to_string
  end
  
  def is_different?(candidate,base) do
    String.downcase(candidate) != String.downcase(base)
  end
end

File: javascript/anagram/anagram.js
var Anagram = function(subject) {
    this.subject = subject;
}

Anagram.prototype.matches = function(matches) {
    var matchesArr = [];
    if (typeof matches == typeof '') {
        matchesArr = Array.from(arguments);
    } else if (typeof matches == typeof []) {
        matchesArr = matches;
    }
    const subject = this.subject;
    function sortIt(word) {
        return word.toLowerCase().split('').sort().join('');
    }
    const sortedSubject = sortIt(subject);
    return matchesArr.reduce( (acc,item) => {
        const sortedItem = sortIt(item);
        const theSame = subject.toLowerCase() == item.toLowerCase();
        if (sortedItem == sortedSubject && !theSame) {
            return acc.concat(item);    
        } else {
            return acc;
        }
    },[]);
}

module.exports = Anagram;
